compiler/optimizer/abstractinterpreter/AbsStackMachineState.cpp

```diff
@@ -0,0 +1,199 @@
+      }
+   return copy;
+   }
+
+TR::AbsValue* TR::AbsOpStack::pop()
+   {
+   TR_ASSERT_FATAL(size() > 0, "Pop an empty stack!");
+   TR::AbsValue *value = _container.back();
+   _container.pop_back();
+   return value;
+   }
+
+void TR::AbsOpStack::merge(TR::AbsOpStack* other)
+   {
+   TR_ASSERT_FATAL(other->_container.size() == _container.size(), "Stacks have different sizes!");
+
+   for (size_t i = 0; i < _container.size(); i ++)
+      {
+      _container[i]->merge(other->_container[i]);
+      }
+   }
+
+void TR::AbsOpStack::setToTop()
+   {
+   for (size_t i = 0; i <  _container.size(); i ++)
+      {
+      _container[i]->setToTop();
+      }
+   }
+
+void TR::AbsOpStack::print(TR::Compilation* comp)
+   {
+   traceMsg(comp, "Contents of Abstract Operand Stack:\n");
+   
+   const int32_t stackSize = size();
+
+   if (stackSize == 0)
+      {
+      traceMsg(comp, "<empty>\n");
+      traceMsg(comp, "\n");
+      return;
+      }
+   
+   traceMsg(comp, "<top>\n");
+
+   for (int32_t i = 0; i < stackSize; i++) 
+      {
+      TR::AbsValue *value = _container[stackSize - i -1 ];
+      traceMsg(comp, "S[%d] = ", stackSize - i - 1);
+      if (value)
+         value->print(comp);
+      traceMsg(comp, "\n");
+      }
+
+   traceMsg(comp, "<bottom>\n");
+   traceMsg(comp, "\n");
+   }
+
+TR::AbsLocalVarArray* TR::AbsLocalVarArray::clone(TR::Region& region)
+   {
+   TR::AbsLocalVarArray* copy = new (region) TR::AbsLocalVarArray(region);
+   for (size_t i = 0; i < _container.size(); i ++)
+      {
+      copy->_container.push_back(_container[i] ? _container[i]->clone(region) : NULL);
+      }
+   return copy;
+   }
+
+void TR::AbsLocalVarArray::setToTop()
+   {
+   for (size_t i = 0; i < _container.size(); i ++)
+      {
+      if (_container[i])
+         _container[i]->setToTop();
+      }
+   }
+
+void TR::AbsLocalVarArray::merge(TR::AbsLocalVarArray* other)
+   {
+   const int32_t otherSize = other->size();
+   const int32_t selfSize = size();
+   const int32_t mergedSize = std::max(selfSize, otherSize);
+
+   for (int32_t i = 0; i < mergedSize; i++)
+      {
+      TR::AbsValue *selfValue = i < size() ? at(i) : NULL;
+      TR::AbsValue *otherValue = i < other->size() ? other->at(i) : NULL;
+
+      if (!selfValue && !otherValue) 
+         {
+         continue;
+         }
+      else if (selfValue && otherValue) 
+         {
+         TR::AbsValue* mergedVal = selfValue->merge(otherValue);
+         set(i, mergedVal);
+         } 
+      else if (selfValue) 
+         {
+         set(i, selfValue);
+         } 
+      else
+         {
+         set(i, otherValue);
+         }
+      }
+   }
+
+void TR::AbsLocalVarArray::set(uint32_t index, TR::AbsValue *value)
+   {
+   if (size() <= index)
+      {
+      _container.resize(index + 1);
+      }
+   _container[index] = value;
+   }
+
+TR::AbsValue* TR::AbsLocalVarArray::at(uint32_t index)
+   {
+   TR_ASSERT_FATAL(index < size(), "Index out of range!");
```

Comment:
The uninitialized, in-bound local variables are NULL. Trying to access an out-of-bound local variable shouldn't happen.

Reply:
So why do we grow the array if we call `set` out-of-bounds then? If out-of-bounds accesses shouldn't happen, then surely we should accept a fixed size during construction and `set` should include this assert as well?

Reply:
Setting a fixed size array is a good idea. I will change to that

Reply:
After some thinking, pre-sizing  the array may not be a good idea since it will add memory footprint. The implementation fo array is growing on demand. So returning NULL for out-of-bound access I think will be a better idea.

Reply:
AbsLocalVarArray is now constructed with a pre-sized `vector`